source/opt/common_uniform_elim_pass.cpp

```diff
@@ -130,6 +130,51 @@ ir::Instruction* CommonUniformElimPass::GetPtr(
   return ptrInst;
 }
 
+bool CommonUniformElimPass::IsVolatileStructType(const ir::Instruction& AccessChainInst) {
+  assert(AccessChainInst.opcode() == SpvOpAccessChain);
+  uint32_t memberTypeId = AccessChainInst.GetSingleWordOperand(3);
+  // OpConstant
+  const ir::Instruction* memberTypeDef = def_use_mgr_->GetDef(memberTypeId);
+  uint32_t member = memberTypeDef->GetSingleWordOperand(2);
+  uint32_t VarId = AccessChainInst.GetSingleWordOperand(2);
+  // OpVariable
+  const ir::Instruction* VarInst = def_use_mgr_->GetDef(VarId);
+  uint32_t ptrTypeId = VarInst->GetSingleWordOperand(0);
+  // OpTypePointer
+  const ir::Instruction* TypePtrInst = def_use_mgr_->GetDef(ptrTypeId);
+  uint32_t typeId = TypePtrInst->GetSingleWordInOperand(kTypePointerTypeIdInIdx);
+  for (auto * dec : dec_mgr_->GetDecorationsFor(typeId, false)) {
+    uint32_t decoration = 0;
+    switch (dec->opcode()) {
+      case SpvOpDecorate:
+        decoration = dec->GetSingleWordOperand(1);
+        break;
+      case SpvOpMemberDecorate:
+        // check if decoration applies to member in OpAccessChain
+        if (member != dec->GetSingleWordOperand(1))
```

Comment:
This code does the right thing if there are two member decorations.

This is an interesting test case.

struct S {
    lowp volatile float x;  
};

I would expect both  
  OpMemberDecorate %S 0 RelaxedPrecision  // possibly, I haven't checked
  OpMemberDecorate %S 0 Volatile

Reply:
I suggest you add a test case like this.  It will ensure the code isn't broken by a rewrite.

Reply:
I tested it and it's fine. It adds just another decoration as expected. I iterate over all decorations, so there shouldn't be a problem.
However, I realized that the we already don't want to optimize if _any_ member is decorated volatile, so i skipped that part.